File: 백준/Silver/20006. 랭킹전 대기열/랭킹전 대기열.java
import java.io.*;
import java.util.*;

public class Main {
    static class Player {
        int level;
        String name;
        Player(int level, String name) {
            this.level = level;
            this.name = name;
        }
    }

    static class Room {
        int leader;
        List<Player> players;
        Room(int leader, Player firstPlayer) {
            this.leader = leader;
            this.players = new ArrayList<>();
            this.players.add(firstPlayer);
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int p = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        List<Player> infos = new ArrayList<>();
        for (int i = 0; i < p; i++) {
            st = new StringTokenizer(br.readLine());
            int l = Integer.parseInt(st.nextToken());
            String n = st.nextToken();
            infos.add(new Player(l, n));
        }

        List<Room> rooms = new ArrayList<>();
        for (Player player: infos) {
            boolean isJoined = false;
            for (Room room : rooms) {
                if (Math.abs(room.leader - player.level) <= 10 && room.players.size() < m) {
                    room.players.add(player);
                    isJoined = true;
                    break;
                }
            }
            if (!isJoined) {
                rooms.add(new Room(player.level, player));
            }
        }

        StringBuilder sb = new StringBuilder();
        for (Room room : rooms) {
            List<Player> sortedPlayers = new ArrayList<>(room.players);
            sortedPlayers.sort(Comparator.comparing(ply -> ply.name));

            if (sortedPlayers.size() == m) {
                sb.append("Started!\n");
            } else {
                sb.append("Waiting!\n");
            }

            for (Player ply : sortedPlayers) {
                sb.append(ply.level).append(" ").append(ply.name).append("\n");
            }
        }

        System.out.print(sb);

    }
}

File: 백준/Silver/20006. 랭킹전 대기열/랭킹전 대기열.py
import sys
input = sys.stdin.readline

p, m = map(int, input().split())
infos = []
for _ in range(p):
    l, n = input().split()
    infos.append([int(l), n])

# for x in infos:
#     print(x)

rooms = []

for l, n in infos:
    if len(rooms) == 0:
        new_room = []
        new_room.append([l, n])
        rooms.append(new_room)
    else:
        is_joined = False
        for i in range(len(rooms)):
            current_room = rooms[i]
            manager_level = current_room[0][0]

            if (abs(manager_level - l) <= 10) and len(current_room) < m:
                current_room.append([l, n])
                is_joined = True
                break
            else: # 다음방 검사
                continue
        if is_joined == False:
            new_room = []
            new_room.append([l, n])
            rooms.append(new_room)


for room in rooms:
    sorted_room = sorted(room, key=lambda x: x[1])
    if len(sorted_room) == m:
        print("Started!")
        for i in range(m):
            print(*sorted_room[i])
    else:
        print("Waiting!")
        for i in range(len(sorted_room)):
            print(*sorted_room[i])




